File: sdk/src/main/java/org/p2p/solanaj/programs/TokenProgram.java
package org.p2p.solanaj.programs;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.math.BigInteger;
import java.util.ArrayList;

import org.p2p.solanaj.core.AbstractData;
import org.p2p.solanaj.core.AccountMeta;
import org.p2p.solanaj.core.PublicKey;
import org.p2p.solanaj.core.TransactionInstruction;
import org.p2p.solanaj.utils.ByteUtils;
import org.p2p.solanaj.core.Sysvar;

public class TokenProgram {
    public static final PublicKey PROGRAM_ID = new PublicKey("TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA");

    public static final int INSTRUCTION_INDEX_INITIALIZE_MINT = 0;
    public static final int INSTRUCTION_INDEX_INITIALIZE_ACCOUNT = 1;
    public static final int INSTRUCTION_INDEX_TRANSFER = 3;
    public static final int INSTRUCTION_INDEX_APPROVE = 4;
    public static final int INSTRUCTION_INDEX_MINT_TO = 7;
    public static final int INSTRUCTION_INDEX_CLOSE_ACCOUNT = 9;

    public static TransactionInstruction initializeMintInstruction(PublicKey tokenProgramId, PublicKey mint,
                                                                   int decimals, PublicKey authority, PublicKey freezeAuthority) {
        ArrayList<AccountMeta> keys = new ArrayList<AccountMeta>();
        keys.add(new AccountMeta(mint, false, true));
        keys.add(new AccountMeta(Sysvar.SYSVAR_RENT_ADDRESS, false, false));

        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        bos.write(INSTRUCTION_INDEX_INITIALIZE_MINT);
        bos.write(decimals);
       // bos.writeBytes(authority.toByteArray());
        bos.write(authority.toByteArray(), 0, authority.toByteArray().length);
        bos.write(freezeAuthority == null ? 0 : 1);
        //bos.writeBytes(freezeAuthority != null ? freezeAuthority.toByteArray() : new byte[PublicKey.PUBLIC_KEY_LENGTH]);
        if (freezeAuthority != null) {
            bos.write(freezeAuthority.toByteArray(), 0, freezeAuthority.toByteArray().length);
        } else {
            byte[] publicKeyArray = new byte[PublicKey.PUBLIC_KEY_LENGTH];
            bos.write(publicKeyArray, 0, publicKeyArray.length);
        }
        return new TransactionInstruction(tokenProgramId, keys, bos.toByteArray());
    }

    public static TransactionInstruction initializeAccountInstruction(PublicKey tokenProgramId, PublicKey account,
                                                                      PublicKey mint, PublicKey owner) {
        ArrayList<AccountMeta> keys = new ArrayList<AccountMeta>();
        keys.add(new AccountMeta(account, false, true));
        keys.add(new AccountMeta(mint, false, false));
        keys.add(new AccountMeta(owner, false, false));
        keys.add(new AccountMeta(Sysvar.SYSVAR_RENT_ADDRESS, false, false));

        byte[] data = new byte[]{INSTRUCTION_INDEX_INITIALIZE_ACCOUNT};

        return new TransactionInstruction(tokenProgramId, keys, data);
    }

    public static TransactionInstruction transferInstruction(PublicKey tokenProgramId, PublicKey source,
                                                             PublicKey destination, PublicKey owner, BigInteger amount) {
        ArrayList<AccountMeta> keys = new ArrayList<AccountMeta>();
        keys.add(new AccountMeta(source, false, true));
        keys.add(new AccountMeta(destination, false, true));
        keys.add(new AccountMeta(owner, true, true));

        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        bos.write(INSTRUCTION_INDEX_TRANSFER);

        try {
            ByteUtils.uint64ToByteStreamLE(amount, bos);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        return new TransactionInstruction(tokenProgramId, keys, bos.toByteArray());
    }

    public static TransactionInstruction approveInstruction(PublicKey tokenProgramId, PublicKey account,
                                                            PublicKey delegate, PublicKey owner, BigInteger amount) {
        ArrayList<AccountMeta> keys = new ArrayList<AccountMeta>();
        keys.add(new AccountMeta(account, false, true));
        keys.add(new AccountMeta(delegate, false, false));
        keys.add(new AccountMeta(owner, true, true));

        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        bos.write(INSTRUCTION_INDEX_APPROVE);

        try {
            ByteUtils.uint64ToByteStreamLE(amount, bos);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        return new TransactionInstruction(tokenProgramId, keys, bos.toByteArray());
    }

    public static TransactionInstruction mintToInstruction(PublicKey tokenProgramId, PublicKey mint,
                                                           PublicKey destination, PublicKey authority, BigInteger amount) {
        ArrayList<AccountMeta> keys = new ArrayList<AccountMeta>();
        keys.add(new AccountMeta(mint, false, true));
        keys.add(new AccountMeta(destination, false, true));
        keys.add(new AccountMeta(authority, true, true));

        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        bos.write(INSTRUCTION_INDEX_MINT_TO);

        try {
            ByteUtils.uint64ToByteStreamLE(amount, bos);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        return new TransactionInstruction(tokenProgramId, keys, bos.toByteArray());
    }

    public static TransactionInstruction closeAccountInstruction(PublicKey tokenProgramId, PublicKey account,
                                                                 PublicKey destination, PublicKey owner) {
        ArrayList<AccountMeta> keys = new ArrayList<AccountMeta>();
        keys.add(new AccountMeta(account, false, true));
        keys.add(new AccountMeta(destination, false, true));
        keys.add(new AccountMeta(owner, false, false));

        byte[] data = new byte[]{INSTRUCTION_INDEX_CLOSE_ACCOUNT};

        return new TransactionInstruction(tokenProgramId, keys, data);
    }

    public static class MintData extends AbstractData {
        public static final int MINT_DATA_LENGTH = ByteUtils.UINT_32_LENGTH + PublicKey.PUBLIC_KEY_LENGTH
                + ByteUtils.UINT_64_LENGTH + 1 + 1 + ByteUtils.UINT_32_LENGTH + PublicKey.PUBLIC_KEY_LENGTH;

        private long mintAuthorityOption;
        private String mintAuthority;
        private BigInteger supply;
        private int decimals;
        private boolean isInitialized;
        private long freezeAuthorityOption;
        private PublicKey freezeAuthority;

        private MintData(byte[] data) {
            super(data, MINT_DATA_LENGTH);

            mintAuthorityOption = readUint32();
            mintAuthority = readPublicKey().toBase58();
            supply = readUint64();
            decimals = readByte();
            isInitialized = readByte() != 0;
            freezeAuthorityOption = readUint32();
            freezeAuthority = readPublicKey();

            if (mintAuthorityOption == 0) {
                mintAuthority = null;
            }

            if (freezeAuthorityOption == 0) {
                freezeAuthority = null;
            }
        }

        public static MintData decode(byte[] data) {
            return new MintData(data);
        }

        public String getMintAuthority() {
            return mintAuthority;
        }

        public PublicKey getMintAuthorityPublicKey() {
            return new PublicKey(mintAuthority);
        }

        public BigInteger getSupply() {
            return supply;
        }

        public int getDecimals() {
            return decimals;
        }

        public boolean isInitialized() {
            return isInitialized;
        }

        public PublicKey getFreezeAuthority() {
            return freezeAuthority;
        }

    }

    public static class AccountInfoData extends AbstractData {
        public static final int ACCOUNT_INFO_DATA_LENGTH = PublicKey.PUBLIC_KEY_LENGTH + PublicKey.PUBLIC_KEY_LENGTH
                + ByteUtils.UINT_64_LENGTH + ByteUtils.UINT_32_LENGTH + PublicKey.PUBLIC_KEY_LENGTH + 1
                + ByteUtils.UINT_32_LENGTH + ByteUtils.UINT_64_LENGTH + ByteUtils.UINT_64_LENGTH
                + ByteUtils.UINT_32_LENGTH + PublicKey.PUBLIC_KEY_LENGTH;

        private PublicKey mint;
        private PublicKey owner;
        private BigInteger amount;
        private long delegateOption;
        private PublicKey delegate;
        private boolean isInitialized;
        private boolean isFrozen;
        private int state;
        private long isNativeOption;
        private BigInteger rentExemptReserve;
        private BigInteger isNativeRaw;
        private boolean isNative;
        private BigInteger delegatedAmount;
        private long closeAuthorityOption;
        private PublicKey closeAuthority;

        private AccountInfoData(byte[] data) {
            super(data, ACCOUNT_INFO_DATA_LENGTH);

            mint = readPublicKey();
            owner = readPublicKey();
            amount = readUint64();
            delegateOption = readUint32();
            delegate = readPublicKey();
            state = readByte();
            isNativeOption = readUint32();
            isNativeRaw = readUint64();
            delegatedAmount = readUint64();
            closeAuthorityOption = readUint32();
            closeAuthority = readPublicKey();

            if (delegateOption == 0) {
                delegate = null;
                delegatedAmount = BigInteger.valueOf(0);
            }

            isInitialized = state != 0;
            isFrozen = state == 2;

            if (isNativeOption == 1) {
                rentExemptReserve = isNativeRaw;
                isNative = true;
            } else {
                rentExemptReserve = null;
                isNative = false;
            }

            if (closeAuthorityOption == 0) {
                closeAuthority = null;
            }
        }

        public static AccountInfoData decode(byte[] data) {
            return new AccountInfoData(data);
        }

        public PublicKey getMint() {
            return mint;
        }

        public PublicKey getOwner() {
            return owner;
        }

        public BigInteger getAmount() {
            return amount;
        }

        public PublicKey getDelegate() {
            return delegate;
        }

        public boolean isNative() {
            return isNative;
        }

        public boolean isInitialized() {
            return isInitialized;
        }

        public boolean isFrozen() {
            return isFrozen;
        }

        public BigInteger getRentExemptReserve() {
            return rentExemptReserve;
        }

        public BigInteger getDelegatedAmount() {
            return delegatedAmount;
        }

        public PublicKey getCloseAuthority() {
            return closeAuthority;
        }
    }

}

File: solanaj/src/main/java/org/p2p/solanaj/programs/TokenProgram.kt
package org.p2p.solanaj.programs

import org.p2p.solanaj.model.core.AbstractData
import org.p2p.solanaj.model.core.AccountMeta
import org.p2p.solanaj.model.core.PublicKey
import org.p2p.solanaj.model.core.Sysvar.SYSVAR_RENT_ADDRESS
import org.p2p.solanaj.model.core.TransactionInstruction
import org.p2p.solanaj.utils.ByteUtils
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.math.BigInteger
import java.util.ArrayList

object TokenProgram {
    val PROGRAM_ID = PublicKey("TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA")
    val ASSOCIATED_TOKEN_PROGRAM_ID = PublicKey("ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL")
    private const val INSTRUCTION_INDEX_INITIALIZE_MINT = 0
    private const val INSTRUCTION_INDEX_INITIALIZE_ACCOUNT = 1
    private const val INSTRUCTION_INDEX_TRANSFER = 3
    private const val INSTRUCTION_INDEX_APPROVE = 4
    private const val INSTRUCTION_INDEX_MINT_TO = 7
    private const val INSTRUCTION_INDEX_CLOSE_ACCOUNT = 9
    private const val INSTRUCTION_INDEX_TRANSFER_CHECKED = 12

    fun createAssociatedTokenAccountInstruction(
        mint: PublicKey,
        associatedAccount: PublicKey,
        owner: PublicKey,
        payer: PublicKey
    ): TransactionInstruction {
        val keys = ArrayList<AccountMeta>()
        keys.add(AccountMeta(payer, isSigner = true, isWritable = true))
        keys.add(AccountMeta(associatedAccount, isSigner = false, isWritable = true))
        keys.add(AccountMeta(owner, isSigner = false, isWritable = false))
        keys.add(AccountMeta(mint, isSigner = false, isWritable = false))
        keys.add(AccountMeta(SystemProgram.PROGRAM_ID, isSigner = false, isWritable = false))
        keys.add(AccountMeta(PROGRAM_ID, isSigner = false, isWritable = false))
        keys.add(AccountMeta(SYSVAR_RENT_ADDRESS, isSigner = false, isWritable = false))
        val data = ByteArray(0)
        return TransactionInstruction(ASSOCIATED_TOKEN_PROGRAM_ID, keys, data)
    }

    fun initializeMintInstruction(
        tokenProgramId: PublicKey,
        mint: PublicKey,
        decimals: Int,
        authority: PublicKey,
        freezeAuthority: PublicKey?
    ): TransactionInstruction {
        val keys = ArrayList<AccountMeta>()
        keys.add(AccountMeta(mint, isSigner = false, isWritable = true))
        keys.add(AccountMeta(SYSVAR_RENT_ADDRESS, isSigner = false, isWritable = false))
        val bos = ByteArrayOutputStream()
        bos.write(INSTRUCTION_INDEX_INITIALIZE_MINT)
        bos.write(decimals)
        // bos.writeBytes(authority.toByteArray());
        bos.write(authority.toByteArray(), 0, authority.toByteArray().size)
        bos.write(if (freezeAuthority == null) 0 else 1)
        // bos.writeBytes(freezeAuthority != null ? freezeAuthority.toByteArray() : new byte[PublicKey.PUBLIC_KEY_LENGTH]);
        if (freezeAuthority != null) {
            bos.write(freezeAuthority.toByteArray(), 0, freezeAuthority.toByteArray().size)
        } else {
            val publicKeyArray = ByteArray(PublicKey.PUBLIC_KEY_LENGTH)
            bos.write(publicKeyArray, 0, publicKeyArray.size)
        }
        return TransactionInstruction(tokenProgramId, keys, bos.toByteArray())
    }

    fun createTransferCheckedInstruction(
        tokenProgramId: PublicKey,
        source: PublicKey,
        mint: PublicKey,
        destination: PublicKey,
        owner: PublicKey,
        amount: BigInteger,
        decimals: Int
    ): TransactionInstruction {
        val keys = ArrayList<AccountMeta>()
        keys.add(AccountMeta(source, isSigner = false, isWritable = true))
        keys.add(AccountMeta(mint, isSigner = false, isWritable = false))
        keys.add(AccountMeta(destination, isSigner = false, isWritable = true))
        keys.add(AccountMeta(owner, isSigner = true, isWritable = false))
        val bos = ByteArrayOutputStream()
        bos.write(INSTRUCTION_INDEX_TRANSFER_CHECKED)
        try {
            ByteUtils.uint64ToByteStreamLE(amount, bos)
        } catch (e: IOException) {
            throw java.lang.RuntimeException(e)
        }
        bos.write(decimals)
        return TransactionInstruction(tokenProgramId, keys, bos.toByteArray())
    }

    fun initializeAccountInstruction(
        tokenProgramId: PublicKey = PROGRAM_ID,
        account: PublicKey,
        mint: PublicKey,
        owner: PublicKey
    ): TransactionInstruction {
        val keys = ArrayList<AccountMeta>()
        keys.add(AccountMeta(account, isSigner = false, isWritable = true))
        keys.add(AccountMeta(mint, isSigner = false, isWritable = false))
        keys.add(AccountMeta(owner, isSigner = false, isWritable = false))
        keys.add(AccountMeta(SYSVAR_RENT_ADDRESS, isSigner = false, isWritable = false))
        val data = byteArrayOf(INSTRUCTION_INDEX_INITIALIZE_ACCOUNT.toByte())
        return TransactionInstruction(tokenProgramId, keys, data)
    }

    fun transferInstruction(
        tokenProgramId: PublicKey,
        source: PublicKey,
        destination: PublicKey,
        owner: PublicKey,
        amount: BigInteger?
    ): TransactionInstruction {
        val keys = ArrayList<AccountMeta>()
        keys.add(AccountMeta(source, isSigner = false, isWritable = true))
        keys.add(AccountMeta(destination, isSigner = false, isWritable = true))
        keys.add(AccountMeta(owner, isSigner = true, isWritable = true))
        val bos = ByteArrayOutputStream()
        bos.write(INSTRUCTION_INDEX_TRANSFER)
        try {
            ByteUtils.uint64ToByteStreamLE(amount, bos)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
        return TransactionInstruction(tokenProgramId, keys, bos.toByteArray())
    }

    fun approveInstruction(
        tokenProgramId: PublicKey,
        account: PublicKey?,
        delegate: PublicKey,
        owner: PublicKey,
        amount: BigInteger
    ): TransactionInstruction {
        val keys = ArrayList<AccountMeta>()
        keys.add(AccountMeta(account!!, isSigner = false, isWritable = true))
        keys.add(AccountMeta(delegate, isSigner = false, isWritable = false))
        keys.add(AccountMeta(owner, isSigner = true, isWritable = false))
        val bos = ByteArrayOutputStream()
        bos.write(INSTRUCTION_INDEX_APPROVE)
        try {
            ByteUtils.uint64ToByteStreamLE(amount, bos)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
        return TransactionInstruction(tokenProgramId, keys, bos.toByteArray())
    }

    fun mintToInstruction(
        tokenProgramId: PublicKey,
        mint: PublicKey,
        destination: PublicKey,
        authority: PublicKey,
        amount: BigInteger
    ): TransactionInstruction {
        val keys = ArrayList<AccountMeta>()
        keys.add(AccountMeta(mint, isSigner = false, isWritable = true))
        keys.add(AccountMeta(destination, isSigner = false, isWritable = true))
        keys.add(AccountMeta(authority, isSigner = true, isWritable = true))
        val bos = ByteArrayOutputStream()
        bos.write(INSTRUCTION_INDEX_MINT_TO)
        try {
            ByteUtils.uint64ToByteStreamLE(amount, bos)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
        return TransactionInstruction(tokenProgramId, keys, bos.toByteArray())
    }

    fun closeAccountInstruction(
        tokenProgramId: PublicKey = PROGRAM_ID,
        account: PublicKey,
        destination: PublicKey,
        owner: PublicKey
    ): TransactionInstruction {
        val keys = ArrayList<AccountMeta>()
        keys.add(AccountMeta(account, isSigner = false, isWritable = true))
        keys.add(AccountMeta(destination, isSigner = false, isWritable = true))
        keys.add(AccountMeta(owner, isSigner = false, isWritable = false))
        val data = byteArrayOf(INSTRUCTION_INDEX_CLOSE_ACCOUNT.toByte())
        return TransactionInstruction(tokenProgramId, keys, data)
    }

    class MintData private constructor(data: ByteArray) : AbstractData(data, MINT_DATA_LENGTH) {

        private val mintAuthorityOption: Long = readUint32()

        var mintAuthority: String?
        val supply: BigInteger
        val decimals: Int
        val isInitialized: Boolean
        private val freezeAuthorityOption: Long
        var freezeAuthority: PublicKey?

        val mintAuthorityPublicKey: PublicKey
            get() = PublicKey(mintAuthority!!)

        companion object {
            const val MINT_DATA_LENGTH = (
                ByteUtils.UINT_32_LENGTH + PublicKey.PUBLIC_KEY_LENGTH +
                    ByteUtils.UINT_64_LENGTH + 1 + 1 + ByteUtils.UINT_32_LENGTH + PublicKey.PUBLIC_KEY_LENGTH
                )

            fun decode(data: ByteArray): MintData {
                return MintData(data)
            }
        }

        init {
            mintAuthority = readPublicKey().toBase58()
            supply = readUint64()
            decimals = readByte().toInt()
            isInitialized = readByte().toInt() != 0
            freezeAuthorityOption = readUint32()
            freezeAuthority = readPublicKey()
            if (mintAuthorityOption == 0L) {
                mintAuthority = null
            }
            if (freezeAuthorityOption == 0L) {
                freezeAuthority = null
            }
        }
    }

    class AccountInfoData private constructor(data: ByteArray) : AbstractData(data, ACCOUNT_INFO_DATA_LENGTH) {
        val mint: PublicKey = readPublicKey()
        val owner: PublicKey = readPublicKey()
        val amount: BigInteger = readUint64()
        private val delegateOption: Long = readUint32()

        companion object {

            const val ACCOUNT_INFO_DATA_LENGTH = (
                PublicKey.PUBLIC_KEY_LENGTH + PublicKey.PUBLIC_KEY_LENGTH +
                    ByteUtils.UINT_64_LENGTH + ByteUtils.UINT_32_LENGTH + PublicKey.PUBLIC_KEY_LENGTH + 1 +
                    ByteUtils.UINT_32_LENGTH + ByteUtils.UINT_64_LENGTH + ByteUtils.UINT_64_LENGTH +
                    ByteUtils.UINT_32_LENGTH + PublicKey.PUBLIC_KEY_LENGTH
                )

            fun decode(data: ByteArray): AccountInfoData {
                return AccountInfoData(data)
            }
        }

        var delegate: PublicKey?
        val isInitialized: Boolean
        val isFrozen: Boolean
        private val state: Int
        private val isNativeOption: Long
        var rentExemptReserve: BigInteger? = null
        private val isNativeRaw: BigInteger
        var isNative = false
        var delegatedAmount: BigInteger
        private val closeAuthorityOption: Long
        var closeAuthority: PublicKey?

        init {
            delegate = readPublicKey()
            state = readByte().toInt()
            isNativeOption = readUint32()
            isNativeRaw = readUint64()
            delegatedAmount = readUint64()
            closeAuthorityOption = readUint32()
            closeAuthority = readPublicKey()
            if (delegateOption == 0L) {
                delegate = null
                delegatedAmount = BigInteger.valueOf(0)
            }
            isInitialized = state != 0
            isFrozen = state == 2
            if (isNativeOption == 1L) {
                rentExemptReserve = isNativeRaw
                isNative = true
            } else {
                rentExemptReserve = null
                isNative = false
            }
            if (closeAuthorityOption == 0L) {
                closeAuthority = null
            }
        }
    }
}